components/butterfly/src/rumor/dat_file.rs

```diff
@@ -49,24 +47,23 @@ pub struct DatFile {
 }
 
 impl DatFile {
-    pub fn new<T: AsRef<Path>>(member_id: &str, data_path: T) -> Self {
-        DatFile { path:        data_path.as_ref().join(format!("{}.rst", member_id)),
+    pub fn new(data_path: PathBuf) -> Self {
+        DatFile { path:        data_path,
                   header_size: 0,
                   header:      Header::default(), }
     }
 
     pub fn path(&self) -> &Path { &self.path }
 
-    pub fn read_into(&mut self, server: &Server) -> Result<()> {
-        let mut version = [0; 1];
-        let mut size_buf = [0; 8];
-        // JW: Resizing this buffer is terrible for performance, but it's the easiest way to
-        // read exactly N bytes from a file. I'm not sure what the right approach is but this
-        // won't be a performance issue for a long time anyway, if ever.
-        let mut rumor_buf: Vec<u8> = vec![];
-        let mut bytes_read = 0;
+    pub fn reader_for_file(&mut self) -> Result<BufReader<File>> {
         let file = File::open(&self.path).map_err(|err| Error::DatFileIO(self.path.clone(), err))?;
-        let mut reader = BufReader::new(file);
+        Ok(BufReader::new(file))
+    }
+
+    pub fn read_header(&mut self,
+                       mut version: &mut [u8],
```

Comment:
Good catch, will update.

Reply:
I'm surprised the compiler didn't complain

Reply:
The reason the compiler didn't complain is because
```rust
        reader.read_exact(&mut version)
```
*does* require `version` to be `mut`; you can't take a `&mut` reference to a non-`mut` binding. The code before was passing `&mut &mut [u8]` to a function that required `&mut [u8]`, but [automatic deref coercion](https://doc.rust-lang.org/book/ch15-02-deref.html#implicit-deref-coercions-with-functions-and-methods) makes that work.

If the call is changed to
```rust
        reader.read_exact(version)
```
the complier should complain about the unnecessary `mut` in the function signature.

Since it can be hard to reason about sometimes, I almost always leave the `mut`s out until the the compiler tells me they're necessary, and then I try to make sure I understand why it's the case.